File: 0x0z0n_Ov3rrid3/0x0z0n-Ov3rride3/src/main.js
// Global Invoke Bridge for Vanilla JS
const invoke = window.__TAURI__.core ? window.__TAURI__.core.invoke : window.__TAURI__.tauri.invoke;

// --- UI ELEMENTS ---
const consoleEl = document.getElementById("console");
const powerSlider = document.getElementById("power-slider");
const powerStatus = document.getElementById("power-status");
const brightSlider = document.getElementById("bright-slider");
const brightStatus = document.getElementById("bright-status");
const appCombo = document.getElementById("app-combo");
const processList = document.getElementById("process-list");
const acWarning = document.getElementById("ac-warning");

const powerModes = [
  { name: "ECO_STEALTH", watts: 45, class: "text-green" },
  { name: "NOMINAL_YIELD", watts: 70, class: "text-cyan" },
  { name: "KINETIC_BURST", watts: 95, class: "text-orange" },
  { name: "OVERRIDE_MAX", watts: 115, class: "text-crimson" }
];

// --- LOGGING ---
function log(msg, isError = false) {
  const color = isError ? "var(--crimson-neon)" : "var(--success-green)";
  consoleEl.innerHTML += `<div style="color: ${color}; margin-bottom: 4px;">> ${msg}</div>`;
  consoleEl.scrollTop = consoleEl.scrollHeight;
}

// --- ENV BUILDER ---
function getEnvPrefix() {
  const useNvidia = document.querySelector('input[name="gpu"]:checked').value === "nvidia";
  let envVars = useNvidia ? "__NV_PRIME_RENDER_OFFLOAD=1 __GLX_VENDOR_LIBRARY_NAME=nvidia __VK_LAYER_NV_optimus=NVIDIA_only " : "";
  if (document.getElementById("chk-gamemode").checked) envVars += "gamemoderun ";
  if (document.getElementById("chk-mangohud").checked) envVars += "mangohud ";
  return envVars;
}

// --- [MODIFIED] 3-TIER WEAPONIZED PROFILES ---
document.getElementById("btn-profile-stealth").addEventListener("click", async () => {
  log("EXECUTING PROFILE: STEALTH");
  powerSlider.value = 0; powerSlider.dispatchEvent(new Event('input')); // ECO_STEALTH (45W)
  brightSlider.value = 3; brightSlider.dispatchEvent(new Event('change'));
  document.querySelector('input[value="integrated"]').checked = true;
  try { await invoke("set_power_limit", { watts: 45 }); } catch (err) { log(err, true); }
});

document.getElementById("btn-profile-light").addEventListener("click", async () => {
  log("EXECUTING PROFILE: LIGHT");
  powerSlider.value = 1; powerSlider.dispatchEvent(new Event('input')); // NOMINAL (70W)
  brightSlider.value = 6; brightSlider.dispatchEvent(new Event('change'));
  document.querySelector('input[value="nvidia"]').checked = true;
  try { await invoke("set_power_limit", { watts: 70 }); } catch (err) { log(err, true); }
});

document.getElementById("btn-profile-boost").addEventListener("click", async () => {
  log("EXECUTING PROFILE: BOOST");
  powerSlider.value = 3; powerSlider.dispatchEvent(new Event('input')); // OVERRIDE (115W)
  brightSlider.value = 10; brightSlider.dispatchEvent(new Event('change'));
  document.querySelector('input[value="nvidia"]').checked = true;
  try {
    await invoke("set_power_limit", { watts: 115 });
    await invoke("spawn_nvidia_terminal");
  } catch (err) { log(err, true); }
});

// --- EXECUTION BUTTONS ---
document.getElementById("btn-browse").addEventListener("click", async () => {
  try {
    const path = await invoke("browse_file");
    appCombo.innerHTML = `<option value="${path}">${path}</option>`;
    log(`TARGET LOCKED: ${path}`);
  } catch (err) { log(err, true); }
});

document.getElementById("btn-launch-gui").addEventListener("click", async () => {
  const target = appCombo.value;
  if (!target || target === "None Selected") return log("ERR: NO BINARY SELECTED", true);
  try {
    const res = await invoke("launch_gui", { target, envVars: getEnvPrefix() });
    log(res);
  } catch (err) { log(`FATAL: ${err}`, true); }
});

document.getElementById("btn-launch-cmd").addEventListener("click", async () => {
  const payloadStr = document.getElementById("cmd-entry").value;
  const payload = payloadStr.trim ? payloadStr.trim() : payloadStr;
  if (!payload) return log("ERR: NO PAYLOAD PROVIDED", true);
  try {
    const res = await invoke("launch_cmd", { payload, envVars: getEnvPrefix() });
    log(res);
  } catch (err) { log(`FATAL: ${err}`, true); }
});

document.getElementById("btn-spawn-shell").addEventListener("click", async () => {
  try {
    log("INITIATING PERMANENT NVIDIA SHELL...");
    const res = await invoke("spawn_nvidia_terminal");
    log(res);
  } catch (err) { log(err, true); }
});

// --- [MODIFIED] SOC MONITOR ---
document.getElementById("btn-soc-monitor").addEventListener("click", async () => {
  try {
    log("LAUNCHING EXTERNAL MONITOR TOOL...");
    const res = await invoke("launch_soc_monitor");
    log(res);
  } catch (err) { log(err, true); }
});

// --- SLIDERS ---
powerSlider.addEventListener("input", (e) => {
  const mode = powerModes[e.target.value];
  powerStatus.innerText = `[ ${mode.name} ] // ${mode.watts}W`;
  powerStatus.className = `status-large font-mono bold ${mode.class}`;
});

document.getElementById("btn-apply-power").addEventListener("click", async () => {
  const watts = powerModes[powerSlider.value].watts;
  log(`INJECTING HARDWARE LIMIT -> ${watts}W`);
  try {
    const res = await invoke("set_power_limit", { watts });
    log(res);
  } catch (err) { log(err, true); }
});

brightSlider.addEventListener("change", async (e) => {
  const level = parseInt(e.target.value) / 10.0;
  brightStatus.innerText = `[ BRIGHTNESS ] // ${e.target.value * 10}%`;
  try { await invoke("set_brightness", { level }); } 
  catch (err) { log(err, true); }
});

// --- [NEW] PROCESS TERMINATOR ---
window.killPid = async function(pid) {
  try { 
    log(`ATTEMPTING TO KILL PID: ${pid}...`);
    const res = await invoke("kill_process", { pid: String(pid) });
    log(res);
  } catch (err) { log(err, true); }
};

// --- [MODIFIED] TELEMETRY LOOP & CORTEX EDR ---
async function updateTelemetry() {
  try {
    // 1. AC State Warning
    const isPlugged = await invoke("get_ac_state");
    acWarning.style.display = isPlugged ? "none" : "block";

    // 2. Hardware Stats
    const data = await invoke("get_telemetry");
    document.getElementById("nv-pct").innerText = `${data.nvidia.usage}%`;
    document.getElementById("nv-bar").style.width = `${data.nvidia.usage}%`;
    document.getElementById("nv-stats").innerText = `${data.nvidia.temp}°C | ${data.nvidia.power.toFixed(1)}W | ${data.nvidia.vram_used}MB`;
    document.getElementById("amd-pct").innerText = `${data.amd.usage}%`;
    document.getElementById("amd-bar").style.width = `${data.amd.usage}%`;

    // 3. Ledger Recording
    if (document.getElementById("chk-ledger").checked) {
      const timestamp = new Date().toISOString();
      const csvLine = `${timestamp},${data.nvidia.temp}C,${data.nvidia.power}W,${data.nvidia.usage}%,${data.nvidia.vram_used}MB`;
      await invoke("append_ledger", { line: csvLine });
    }

    // 4. CORTEX EDR PROCESS RENDERER (Dual-Line Kernel Extraction)
    const procs = await invoke("get_gpu_processes");
    if (procs.length === 0) {
      processList.innerHTML = `<div class="muted small-text font-mono" style="text-align:center; padding-top: 35px; letter-spacing: 1px;">[ CORTEX ] ZERO ACTIVE PAYLOADS DETECTED.</div>`;
    } else {
      let html = `
        <div style="display: flex; justify-content: space-between; font-size: 9px; color: var(--muted-text); font-family: var(--font-mono); padding-bottom: 6px; border-bottom: 1px dashed var(--border-color); margin-bottom: 6px; text-transform: uppercase; letter-spacing: 1px;">
          <span style="width: 50px;">PID</span>
          <span class="flex-grow">PAYLOAD_SIGNATURE / EXECUTION_PATH</span>
          <span style="width: 70px; text-align: right; margin-right: 15px;">FOOTPRINT</span>
          <span style="width: 40px; text-align: center;">ACTION</span>
        </div>
      `;
      
      html += procs.map(p => {
        let memVal = parseInt(p.memory);
        // THREAT LOGIC: > 2GB = Red, > 500MB = Orange, Normal = Cyan
        let threatColor = memVal > 2000 ? 'var(--crimson-neon)' : (memVal > 500 ? 'var(--warning-orange)' : 'var(--cyan-neon)');
        
        return `
        <div style="display: flex; justify-content: space-between; align-items: center; padding: 6px 0; border-bottom: 1px solid rgba(255,255,255,0.03); font-size: 11px;">
          <span class="font-mono bold" style="width: 50px; color: ${threatColor};">${p.pid}</span>
          
          <div class="flex-grow" style="display: flex; flex-direction: column; overflow: hidden; margin-right: 10px;">
            <span class="text-main font-mono bold" style="white-space: nowrap; overflow: hidden; text-overflow: ellipsis;">${p.name}</span>
            <span class="muted font-mono" style="font-size: 9px; white-space: nowrap; overflow: hidden; text-overflow: ellipsis; margin-top: 2px;" title="${p.full_cmd}">${p.full_cmd}</span>
          </div>

          <span class="font-mono bold" style="width: 70px; text-align: right; margin-right: 15px; color: ${threatColor};">${p.memory} MiB</span>
          <button class="btn-kill" style="width: 40px; border: 1px solid ${threatColor}; color: ${threatColor};" onmouseover="this.style.background='${threatColor}'; this.style.color='#000';" onmouseout="this.style.background='transparent'; this.style.color='${threatColor}';" onclick="killPid(${p.pid})">KILL</button>
        </div>
      `}).join('');
      processList.innerHTML = html;
    }
  } catch (err) { 
    processList.innerHTML = `<div class="text-crimson bold font-mono" style="text-align:center; padding-top: 10px;">[!] BACKEND DESYNC: ${err}</div>`;
  }
}

// --- [NEW] GLOBAL HOTKEYS ---
window.addEventListener('keydown', async (e) => {
  if (e.ctrlKey && e.shiftKey && e.code === 'KeyO') {
    log("HOTKEY DETECTED: Spawning Persistent Shell...");
    try { await invoke("spawn_nvidia_terminal"); } catch(err) { log(err, true); }
  }
  if (e.ctrlKey && e.shiftKey && e.code === 'KeyP') {
    log("HOTKEY DETECTED: Toggling Hardware Limit...");
    powerSlider.value = powerSlider.value == 0 ? 3 : 0;
    powerSlider.dispatchEvent(new Event('input'));
    document.getElementById("btn-apply-power").click();
  }
});

log("SYS.CORE INITIATED // TACTICAL MODULES ACTIVE");
setInterval(updateTelemetry, 1000);

File: 0x0z0n_Ov3rrid3/0x0z0n-Ov3rride3/src/main.rs
#![cfg_attr(not(debug_assertions), windows_subsystem = "windows")]

use std::process::Command;
use std::fs;

// [01] TELEMETRY
#[tauri::command]
fn get_telemetry() -> Result<serde_json::Value, String> {
    let nv_output = Command::new("nvidia-smi")
        .args(["--query-gpu=utilization.gpu,memory.used,memory.total,temperature.gpu,power.draw", "--format=csv,noheader,nounits"])
        .output();
    
    let mut nv_data = serde_json::json!({"usage": 0, "vram_used": 0, "vram_total": 0, "temp": 0, "power": 0.0});
    
    if let Ok(output) = nv_output {
        if output.status.success() {
            let stdout = String::from_utf8_lossy(&output.stdout);
            let parts: Vec<&str> = stdout.trim().split(", ").collect();
            if parts.len() == 5 {
                nv_data = serde_json::json!({
                    "usage": parts[0].parse::<u32>().unwrap_or(0),
                    "vram_used": parts[1].parse::<u32>().unwrap_or(0),
                    "vram_total": parts[2].parse::<u32>().unwrap_or(0),
                    "temp": parts[3].parse::<u32>().unwrap_or(0),
                    "power": parts[4].parse::<f32>().unwrap_or(0.0),
                });
            }
        }
    }

    let mut amd_usage = 0;
    let paths = ["/sys/class/drm/card0/device/gpu_busy_percent", "/sys/class/drm/card1/device/gpu_busy_percent"];
    for path in paths {
        if let Ok(content) = fs::read_to_string(path) {
            if let Ok(usage) = content.trim().parse::<u32>() { amd_usage = usage; break; }
        }
    }

    Ok(serde_json::json!({ "nvidia": nv_data, "amd": {"usage": amd_usage} }))
}

// [02] POWER LIMITER
#[tauri::command]
fn set_power_limit(watts: u32) -> Result<String, String> {
    let output = Command::new("pkexec")
        .args(["nvidia-smi", "-pl", &watts.to_string()]).output().map_err(|e| e.to_string())?;
    if output.status.success() { Ok(format!("SUCCESS: TGP LOCKED AT {}W", watts)) } 
    else { Err(format!("ERR: {}", String::from_utf8_lossy(&output.stderr).trim())) }
}

// [03] DISPLAY OVERRIDE
#[tauri::command]
fn set_brightness(level: f32) -> Result<String, String> {
    let xrandr_out = Command::new("sh").arg("-c").arg("xrandr | grep ' connected'").output().map_err(|e| e.to_string())?;
    let stdout = String::from_utf8_lossy(&xrandr_out.stdout);
    let display = stdout.split_whitespace().next().unwrap_or("eDP");

    let output = Command::new("xrandr")
        .args(["--output", display, "--brightness", &level.to_string()]).output().map_err(|e| e.to_string())?;
    if output.status.success() { Ok(format!("SUCCESS: BRIGHTNESS SET TO {}%", (level * 100.0) as i32)) } 
    else { Err("ERR: XRANDR INJECTION FAILED".to_string()) }
}

// [04] FILE BROWSER
#[tauri::command]
fn browse_file() -> Result<String, String> {
    let output = Command::new("zenity")
        .args(["--file-selection", "--title=z0n // SELECT TARGET BINARY", "--filename=/usr/bin/", "--modal"])
        .output().map_err(|e| e.to_string())?;
    if output.status.success() { Ok(String::from_utf8_lossy(&output.stdout).trim().to_string()) } 
    else { Err("BROWSE ABORTED".to_string()) }
}

// [05] GUI LAUNCHER
#[tauri::command]
fn launch_gui(target: &str, env_vars: &str) -> Result<String, String> {
    let cmd = format!("{} '{}'", env_vars, target);
    Command::new("bash").arg("-c").arg(&cmd).spawn().map_err(|e| e.to_string())?;
    Ok(format!("SUCCESS: SPAWNED {}", target))
}

// [06] TERMINAL LAUNCHER
#[tauri::command]
fn launch_cmd(payload: &str, env_vars: &str) -> Result<String, String> {
    let inner_cmd = format!("{}{}", env_vars, payload);
    let terminal_cmd = format!("x-terminal-emulator -e bash -c \"{}; echo; echo '>> SESSION TERMINATED'; exec bash\"", inner_cmd);
    Command::new("bash").arg("-c").arg(&terminal_cmd).spawn().map_err(|e| e.to_string())?;
    Ok("SUCCESS: TERMINAL HOOK ESTABLISHED".to_string())
}

// CORE BUILDER
fn main() {
    tauri::Builder::default()
        .invoke_handler(tauri::generate_handler![
            get_telemetry, 
            set_power_limit, 
            set_brightness, 
            browse_file, 
            launch_gui, 
            launch_cmd
        ])
        .run(tauri::generate_context!())
        .expect("error while running tauri application");
}
